agency/lock.go

```diff
@@ -0,0 +1,204 @@
+// limitations under the License.
+//
+// Copyright holder is ArangoDB GmbH, Cologne, Germany
+//
+// Author Ewout Prangsma
+//
+
+package agency
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"sync"
+	"time"
+
+	driver "github.com/arangodb/go-driver"
+)
+
+const (
+	minLockTTL = time.Second * 5
+)
+
+// Lock is an agency backed exclusive lock.
+type Lock interface {
+	// Lock tries to lock the lock.
+	// If it is not possible to lock, an error is returned.
+	// If the lock is already held by me, an error is returned.
+	Lock(ctx context.Context) error
+
+	// Unlock tries to unlock the lock.
+	// If it is not possible to unlock, an error is returned.
+	// If the lock is not held by me, an error is returned.
+	Unlock(ctx context.Context) error
+
+	// IsLocked return true if the lock is held by me.
+	IsLocked() bool
+}
+
+// Logger abstracts a logger.
+type Logger interface {
+	Errorf(msg string, args ...interface{})
+}
+
+// NewLock creates a new lock on the given key.
+func NewLock(log Logger, api Agency, key []string, id string, ttl time.Duration) (Lock, error) {
+	if ttl < minLockTTL {
+		ttl = minLockTTL
+	}
+	if id == "" {
+		randBytes := make([]byte, 16)
+		rand.Read(randBytes)
+		id = hex.EncodeToString(randBytes)
+	}
+	return &lock{
+		log: log,
+		api: api,
+		key: key,
+		id:  id,
+		ttl: ttl,
+	}, nil
+}
+
+type lock struct {
+	mutex         sync.Mutex
+	log           Logger
+	api           Agency
+	key           []string
+	id            string
+	ttl           time.Duration
+	locked        bool
+	cancelRenewal func()
+}
+
+// Lock tries to lock the lock.
+// If it is not possible to lock, an error is returned.
+// If the lock is already held by me, an error is returned.
+func (l *lock) Lock(ctx context.Context) error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	if l.locked {
+		return driver.WithStack(AlreadyLockedError)
+	}
+
+	// Try to claim lock
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	if err := l.api.WriteKeyIfEmpty(ctx, l.key, l.id, l.ttl); err != nil {
+		if driver.IsPreconditionFailed(err) {
+			return driver.WithStack(AlreadyLockedError)
+		}
+		return driver.WithStack(err)
+	}
+
+	// Success
+	l.locked = true
+
+	// Keep renewing
+	renewCtx, renewCancel := context.WithCancel(context.Background())
+	go l.renewLock(renewCtx)
+	l.cancelRenewal = renewCancel
+
+	return nil
+}
+
+// Unlock tries to unlock the lock.
+// If it is not possible to unlock, an error is returned.
+// If the lock is not held by me, an error is returned.
+func (l *lock) Unlock(ctx context.Context) error {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	if !l.locked {
+		return driver.WithStack(NotLockedError)
+	}
+
+	// Release the lock
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	if err := l.api.RemoveKeyIfEqualTo(ctx, l.key, l.id); err != nil {
+		return driver.WithStack(err)
+	}
+
+	// Cleanup
+	l.locked = false
```

Comment:
I think here is a data race: You first remove the key in the agency, then set `locked` to false (irrelevant here), and only then stop the renewal. I suggest to first cancel the renewal, make sure it is cancelled (unless `l.cancelRenewal()` does this automatically), then remove the key and only then set `locked` to `false`.

Reply:
OK, now I understand, the renewal has a precondition as well. OK, this is good. Nevertheless, I think the change of order makes the code easier understandable. Furthermore, If the `RemoveKeyIfEqualTo` fails, then the renewal go routine might keep running. If we have cancelled first, it is gone in any case.

Reply:
Note that this function (and the renewal) first claim an in-memory mutex.
So there can be no race.

Second. If RemoveKeyIfEqualTo fails, the lock is likely still held, so I expect the renewal to continue.

Reply:
Agreed.